R/tostring.R

```diff
@@ -130,33 +186,83 @@ setMethod("toString", "MatrixPrintForm", function(x,
+  # NB: what about \n in topleft? -> not supported
+  header_indent <- gsub("^([[:space:]]*).*", "\\1", mat$strings[1:nlh, 1]) # Supposedly never with empty strings " "
+  old_indent <- c(header_indent, old_indent)
   need_reindent <- nzchar(old_indent)
-  reindent_old_idx <- mf_lgrouping(mat)[need_reindent]
+  # Check for which row has indent
+  ind_from_strings <- nchar(old_indent)[-seq_len(nlh)] > 0
+  if (!checkmate::test_set_equal(ind_from_strings, ind_from_mf)) {
+    stop("Row-info and string indentations are different.", # nocov
+         " Please contact the maintainer, this should not happen.") # nocov
+  }
+  ori_mflg <- mf_lgrouping(mat) # Original groups
+  reindent_old_idx <- ori_mflg[need_reindent] # Indent groups bf wrap
+
+  # Taking care in advance of indented word wrappings
   cell_widths_mat[need_reindent, 1] <- cell_widths_mat[need_reindent, 1] - nchar(old_indent)[need_reindent]
 
+  # Case in which the indentation is taking too much space vs desired wrapping
+  if (any(cell_widths_mat < 0)) {
+    col_culprits <- apply(cell_widths_mat, 2, function(i) any(i < 0))
+    stop("Inserted width(s) for column(s) ", which(col_culprits),
+         " is(are) not wide enough for the desired indentation.")
+  }
+
   new_strings <- matrix(unlist(mapply(wrap_string,
                                       str = mat$strings,
                                       max_width = cell_widths_mat,
                                       hard = TRUE)),
     ncol = ncol(mat$strings)
   )
   mat$strings <- new_strings
+
   ## XXXXX this is wrong and will break for listings cause we don't know when
   ## we need has_topleft to be FALSE!!!!!!!!!!
   mat <- mform_handle_newlines(mat)
 
   # Adding again the indentation
+  # Indent groups after newline
   reindent_new_idx <- mf_lgrouping(mat) %in% reindent_old_idx
   if (anyNA(reindent_new_idx)) {
     stop(
       "Unable to remap indenting after cell content text wrapping. ", # nocov
-      "Please contact the maintainer, this should not happen"
+      "Please contact the maintainer, this should not happen."
     ) # nocov
   }
+
   # Adding the indentation back in
-  new_indent <- old_indent[mf_lgrouping(mat)[reindent_new_idx]]
-  mat$strings[which(reindent_new_idx), 1] <- paste0(
+  ind_v <- NULL
+  for (i in mf_lgrouping(mat)[reindent_new_idx]) {
+    ind_v <- c(ind_v, which(i == ori_mflg)[1])
+  }
+  new_indent <- old_indent[ind_v]
+
+  # Additional safety check
+  if (length(new_indent) > 0 && !all(nzchar(new_indent))) {
+    stop("Recovered indentation contains empty strings. This is an", # nocov
+         " indexing problem, please contact the maintainer, this should not happen.") # nocov
+  }
+
+  # Indentation is different for topleft material
+  if (isTRUE(mf_has_topleft(mat))) {
+    # mf_nlheader counts actual header lines while mf_nrheader is 'virtual'
+    # A bit of an hack, we should see what happens if there are \n in topleft
```

Comment:
\n is _never_ allowed in top-left, and isn't going to be. They can specify each single line topleft part separately as a vector but there should not ever be new lines there. 

The only timme I can think of that its even possible that could happen is if newlines are getting variable labels and the variable labels have newlines in them, but we can just tell them not to do that, or hard split it in the actual topleft setter. By the time something is in the topleft slot it should *not* have newlines.

Reply:
Yes, I fully agree. Having \n there makes things very complicated without any actual gain. I do not envision people starting to do this too soon. We should add an error for it (another issue I guess).

Reply:
Modified comments accordingly